src/charm.py

```diff
@@ -87,6 +92,115 @@ def _spring_boot_port(self) -> int:
             return port
         return 8080
 
+    def _exec(
+        self, command: typing.List[str], environment: typing.Optional[typing.Dict[str, str]] = None
+    ) -> typing.Tuple[int, str, str]:
+        """Execute a command in Spring Boot application container with a timeout of 60s.
+
+        Args:
+            command: command to be executed.
+            environment: environment variables for the process.
+
+        Returns:
+            A 3-tuple of exit code, stdout and stderr.
+        """
+        container = self._spring_boot_container()
+        process = container.exec(command, environment=environment, timeout=60)
+        try:
+            stdout, stderr = process.wait_output()
+            return 0, stdout, stderr
+        except ExecError as error:
+            return error.exit_code, typing.cast(str, error.stdout), typing.cast(str, error.stderr)
```

Comment:
what about `str(error.stdout)` and `str(typing.stderr)`?

Reply:
The type definition of stdout and stderr as returned by `wait_output` is `None | bytes | str`. Using `str` here is a little bit misleading, since `str(bytes)` and `str(None)` is not valid. `typing.cast` here is to narrow the type, `typing.cast` does nothing in the runtime.

Reply:
Maybe update the type hint to be `str | bytes | None`? Then we don't have to cast and the type hints is more accurate? E.g., users of the function would know they need to handle those cases

Reply:
The `stdout` and `stderr` are always `str` here, since we invoke the `ops.model.Container.exec` with default `encoding` argument, which is `utf-8`.